source/aws_imds_client.c

```diff
@@ -607,15 +615,50 @@ static void s_query_complete(struct imds_user_data *user_data) {
         return;
     }
 
-    /* In this case we fallback to the secure imds flow. */
     if (user_data->status_code == AWS_HTTP_STATUS_CODE_401_UNAUTHORIZED) {
-        s_invalidate_cached_token_safely(user_data);
-        s_reset_scratch_user_data(user_data);
-        aws_retry_token_release(user_data->retry_token);
-        if (s_get_resource_async_with_imds_token(user_data)) {
-            s_user_data_release(user_data);
+        /* In this case we fail the request and switch to secure path for following requests. */
+        if (!user_data->imds_token_required) {
+            if (user_data->is_insecure_fallback) {
+                /* V2 request, but failed to fetch token and fallback failed */
+                AWS_LOGF_ERROR(
+                    AWS_LS_IMDS_CLIENT,
+                    "(id=%p) IMDS client failed to fetch resource through fallback to insecure path. requester %p.",
+                    (void *)user_data->client,
+                    (void *)user_data);
+                user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_INVALID;
+            } else {
+                /* V1 request, V2 required */
+                AWS_LOGF_ERROR(
+                    AWS_LS_IMDS_CLIENT,
+                    "(id=%p) IMDS client failed to fetch resource without token. requester %p.",
+                    (void *)user_data->client,
+                    (void *)user_data);
+                user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_REQUIRED;
+            }
+        } else {
+            /* Not expected, the token was told as invalid from server */
+            AWS_LOGF_ERROR(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client failed to fetch resource. Server response 401 UNAUTHORIZED indicates invalid "
+                "token. requester %p.",
+                (void *)user_data->client,
+                (void *)user_data);
+            user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_INVALID;
```

Comment:
If it is not expected, maybe an assert?

Reply:
I feel like it's still possible that server starts to don't like our cache token for some other reasons, it's probably not worth to crash the program